[Warlock] Initial implementation of affliction T28 set

Fatherdotter opened the pull request:
Working implementation of affliction's T28 2-piece and 4-piece set bonus.

PR is available for review, as discussed we are awaiting PTR testing to confirm behavior before merging.

CurtisJMC committed:
Implemented offspec corruption as an action. Call from the main warlock file does no direct damage and is actioned only for demo/destro, otherwise is passed down to aff as normal.

CurtisJMC committed:
Merge branch 'shadowlands' of https://github.com/Fatherdotter/simc into shadowlands

* 'shadowlands' of https://github.com/Fatherdotter/simc:
  [Druid] simplify name concatenation function
  [Action] allow spells with infinite duration to be used for apply_affecting_aura()
  squash warnings
  [Druid] use static for name concatenation function

CurtisJMC committed:
Add destro mastery to Drain Life/Corruption

Fatherdotter committed:
Merge branch 'simulationcraft:shadowlands' into shadowlands

CurtisJMC committed:
Merge branch 'shadowlands' of https://github.com/Fatherdotter/simc into shadowlands

* 'shadowlands' of https://github.com/Fatherdotter/simc: (48 commits)
  [Warlock] Don't be bad about memory management (#6237)
  [Warlock] Update create_action logic to handle cross-spec conflicts (#6236)
  [Shaman] Less crashing
  Maybe put method overrides in the correct place.
  [Shaman] Add Raging Vesper Vortex as a child of Vesper Totem
  [Shaman] Don't trigger Raging Vesper Vortex without enemies
  [Shaman] More Vesper Totem & Raging Vesper Vortex reimplementation
  Implement action_e string helper
  [Racial] Implement Gift of the Naaru
  [Shaman] Refactor Vesper Totem functionality
  [Shaman] Refactor totem code to support non-damaging totems
  [Warlock] Fixes for Malefic Rapture (#6233)
  [Mage] resolve/adjust some TODOs
  [Druid] use execute_on_target
  [Mage] Minor consistency fixes
  [Mage] switch to execute_on_target
  [Action] add action_t::execute_on_target
  [Mage] Improve formatting
  [Mage] Simplify welly freeze
  Add newline to end of error ouptput
  ...

CurtisJMC committed:
Remove spec check from corruption action in general create_action_warlock()

CurtisJMC committed:
[Warlock] Remove destro mastery implementation from Drain life, in-game testing shows it is not affected

CurtisJMC committed:
[Warlock] Add non-aff periodic spells to WoC whitelist. Testing confirms they can stack the buff via corruption, and it does buff their periodic damage spells.

Fatherdotter committed:
Merge branch 'simulationcraft:shadowlands' into shadowlands

CurtisJMC committed:
[Warlock] Added Sacrolash to the main corruption_t definition, as Demo/Destro can benefit from it.

CurtisJMC committed:
Merge branch 'shadowlands' of https://github.com/Fatherdotter/simc into shadowlands

* 'shadowlands' of https://github.com/Fatherdotter/simc:
  [HTML] correctly display overrides for ptr mode
  [Druid] organize options
  [Feral] update covenant evaluations by experts

CurtisJMC committed:
[Warlock] Remove unnecessary tick() call in main class corruption_t

CurtisJMC committed:
[Warlock] Initial implementation of aff T28 set bonuses. Includes debug variable to demonstrate the current working implementation (actor flags currently not functional)

CurtisJMC committed:
[Warlock] Reverse aff set bonuses, query spell data for calamitous buff chance.

CurtisJMC committed:
[Warlock] Convert dot_extension for 2pc aff bonus from millis to seconds.
Debug variables no longer required.

Fatherdotter committed:
Merge branch 'simulationcraft:shadowlands' into shadowlands

Fatherdotter committed:
Merge branch 'simulationcraft:shadowlands' into warlock_tier

CurtisJMC committed:
Merge branch 'warlock_tier' of https://github.com/Fatherdotter/simc into warlock_tier

CurtisJMC committed:
Merge branch 'shadowlands' of https://github.com/Fatherdotter/simc into warlock_tier

CurtisJMC committed:
[Warlock] Correct typo in Drain Soul proc rate for Calamitous Crescendo.

Azevara commented:
This should be safe to merge following sanity checks once PTR is testable on Dec 2.

scamille commented on engine/class_modules/warlock/sc_warlock_affliction.cpp:
```diff
@@ -106,10 +106,28 @@ struct shadow_bolt_t : public affliction_spell_t
+      if ( p()->sets->has_set_bonus( WARLOCK_AFFLICTION, T28, B4 ) )
+      {        
+        auto tdata = this->td( s->target );
+        bool tierDotsActive = tdata->dots_agony->is_ticking();
```
I would express this with a single boolean statement instead of (mis)using the bitwise operator.

Fatherdotter replied:
Would this be functionally different from the existing implementation in any way? As far as I am aware the usage of 

```c++
bool val = A;
val &= B;
val &= C;
```

Is functionally equivalent to 

```c++
bool val = A && B && C;
```

With the exceptions that they do not short-circuit like a single logical statement would, and thus the minor overhead of guaranteed evaluation of both B and C in exchange for (in my opinion) clarity of the statements to be evaluated in this case. Considering I have recently had yourself and an existing simc member speak against using a similar approach in this project I am not sure the reason for this change, but welcome any advice to the contrary.

Edit: Clarified example.

vituscze replied:
Problem here isn't functionality but meaning. `&` is a bitwise operation (and so is `&=`) but you want a logical operation. Sadly C++ doesn't have logical version of `&=` such as `&&=`, so you're left with either `val = A && B;` or `val = A; val = val && B;`.

By the way, those two are not the same when dealing with expressions that can be converted to `bool`:
```cpp
bool b = true;
b &= 2;
// b is false

bool c = true;
c = c && 2;
// c is true
```

Fatherdotter replied:
I am aware of the bitwise nature of the assignment operators here but did not see how it was functionally different in the event that we are working purely with boolean data types. Since all of these operators are boolean in nature ( is_ticking() ) and would have no reason to become non-boolean values in the future, this interaction should not possibly arise. The only difference in functionality is the short-circuit nature being lost in the expression.

vituscze replied:
It's the same reason we don't use `+` or `*` on booleans (even though it technically works as disjunction/conjunction when all the arguments are booleans): it's just the wrong operator to use.

Fatherdotter replied:
In the absence of a logical assignment operator for c++, for someone who prefers the clarity of having the conditions clearly spelled out there is no other functional alternative though.

I understand the concern here, and although I cannot conceive of a situation where this code fails the change can be made simply to meet conventions at the expense of clarity.

scamille commented on engine/class_modules/warlock/sc_warlock_affliction.cpp:
```diff
@@ -606,6 +650,19 @@ struct malefic_rapture_t : public affliction_spell_t
+        timespan_t dot_extension =  p()->sets->set( WARLOCK_AFFLICTION, T28, B2 )->effectN( 2 ).time_value() * 1000;
+        warlock_td_t* td = p()->get_target_data( target );
+
+        if ( td )
```
the pointer returned by get_target_data should always be non-null.

Fatherdotter replied:
Considering you have, quite recently, argued against saving multiple checks per-action as being an unnecessary optimization surely a single sanity check to prevent dereferencing a null pointer is a reasonable addition to the program?

There is existing similar usage for this exact purpose in simc, and it should add negligible overhead.

vituscze replied:
If `get_target_data` returns a nullptr, something is wrong and it'd be better to crash than to silently ignore it.

Fatherdotter replied:
If that is the decision made here, then I will happily do a separate commit to remove these checks in whole from the warlock modules (rather than removing one instance in this commit, and others in a separate commit.)

scamille replied:
I don't have anything against checking it per se, but would then change it to an assert or throw instead of silently doing nothing.

Fatherdotter replied:
Resolved by the most recent commit. After discussing, we have elected to handle it within this commit and not take the route of an assert or throw.

CurtisJMC committed:
[Warlock] Remove bitwise assignment operators as requested.

CurtisJMC committed:
[Warlock] Remove bitwise assignment operators as requested.

CurtisJMC committed:
[Warlock] Correct typo in recent bitwise assignment operator commit.

CurtisJMC committed:
[Warlock] Remove td null reference check, as requested.

CurtisJMC committed:
[Warlock] T4 Affliction corrections
- Move calamitous_crescendo code out of
   `malefic_rapture_damage_instance_t` and into `malefic_rapture_t`
- Add TODO about hardcoded duration (required for current testing)
   and incorrect spell ID for calamitous_crescendo

CurtisJMC committed:
[Warlock] Correct Aff T28 4pc buff
- Replace stand-in spell ID with the buff ID
- Remove hardcoded duration
- Replace spell queries with set queries for setting proc chance

CurtisJMC committed:
[Warlock] Correction in make_buff for Aff T28 4pc
We cannot reference p() in make_buff, but we can access sets through
  a reference to `this`

CurtisJMC committed:
[Warlock] Remove TODO for T28 Shadow Bolt/Drain Soul interaction
Procs confirmed to occur on damage for affliction T28 set bonus.

vituscze commented on engine/class_modules/warlock/sc_warlock_affliction.cpp:
```diff
@@ -935,7 +934,9 @@ void warlock_t::create_buffs_affliction()
 
+  // TODO: Replace hardcoded duration with query from spell data, replace buff spell (363953) with actuall buff spell ID
   buffs.calamitous_crescendo = make_buff( this, "calamitous_crescendo", find_spell( 363953 ) )
+                                   ->set_duration(timespan_t::from_seconds(10))
```
If you want, you can use `10_s` instead of `timespan_t::from_seconds( 10 )`

Fatherdotter replied:
Appreciate the advice, in the future I will try to put it to use but I do not intend to set these values.

vituscze commented on engine/class_modules/warlock/sc_warlock_affliction.cpp:
```diff
@@ -935,8 +935,8 @@ void warlock_t::create_buffs_affliction()
   buffs.calamitous_crescendo = make_buff( this, "calamitous_crescendo", find_spell( 364322 ) )
-                                   ->set_chance( talents.drain_soul->ok() ? p()->sets->set(WARLOCK_AFFLICTION, T28, B4 )->effectN( 2 ).percent() 
-                                                                          : p()->sets->set(WARLOCK_AFFLICTION, T28, B4 )->effectN( 1 ).percent() );
+                                   ->set_chance( talents.drain_soul->ok() ? this->sets->set(WARLOCK_AFFLICTION, T28, B4 )->effectN( 2 ).percent() 
```
You can drop `this` and just use `sets->set(...)`

Fatherdotter replied:
Thank you for the advice. Dropped this from a recent commit as setting the proc chance was actually a breaking change.

CurtisJMC committed:
[Warlock] Correction to Aff T28 4pc buff
- Remove chance set on calamitous_crescendo buff creation
